### PR TITLE
Add cython data file reader

### PR DESCRIPTION
**Pull request by kevinsa5**
Following up on #6, here is a cython-based data file reader.  I didn't rewrite the entire thing in cython, only the parts at the top that were particularly slow in pandas (reading the CSV and the subsequent string operations). 

I made it so that `setup.py` defaults to trying to compile the C-extension when installing gemlog but made it possible to build without the extension by setting an environment variable so that people without a C compiler can still use the rest of the package.  I chose an env var over a command line flag because getting options like `pip install ... --no-cython` to get passed through to `setup.py` is somewhat difficult.

Presumably the function name `read_with_cython` should be something else, but I figured API decisions are best left to @ajakef.  One idea is creating a new wrapper function that switches between the three reader functions like `read_datafile(filename, method='cython')` where method can also be `'python'` for the original or `'pandas'` for the pandas version?

Haven't really done any testing other than comparing against the original reader function on my three test files:

```python
In [1]: import gemlog
   ...: filename = '/home/kevin/Downloads/FILE0003.077'

In [2]: %time result_fast = gemlog.read_with_cython(filename)
CPU times: user 316 ms, sys: 19.6 ms, total: 335 ms
Wall time: 334 ms

In [3]: %time result_slow = gemlog.slow_ReadGem_v0_9_single(filename, 0)
CPU times: user 6.63 s, sys: 5.71 ms, total: 6.64 s
Wall time: 6.63 s

In [4]: import numpy as np
   ...: import pandas as pd

In [5]: assert np.allclose(result_slow['data'], result_fast['data'])

In [6]: old = result_slow['metadata']
   ...: new = result_fast['metadata'].drop(columns=['index'])
   ...: for column in old:
   ...:     pd.testing.assert_series_equal(old[column], new[column])
   ...: 

In [7]: old = result_slow['gps']
   ...: new = result_fast['gps'].drop(columns=['index'])
   ...: for column in old:
   ...:     pd.testing.assert_series_equal(old[column], new[column])
   ...: 
```

### CODE REVIEW

**Comment by ajakef** [comment]
Thanks for speeding up this function yet again! I like the wrapper function idea.

Just added new tests for edge. The main change relevant to this issue is that ReadGem_v0_9 must raise the EmptyRawFile exception when reading an empty file. This probably introduced a conflict between our branches related to error handling.

**Comment by kanderso-nrel** [comment]
Merge conflict resolved.  A few questions:

1) Right now I've just copied the basic functionality test of `ReadGem_v0_9_single` for this.  Should I copy the other tests as well (empty/corrupt)?
2) For the `read_datafile(filename, method=...)` idea, does it make sense to have a wrapper function at the single datafile level, or would it be better to add that parameter to `ReadGem` and have it switch between the lower level functions directly?
3) Still need to rename `read_cython` to something better -- any preference for a naming scheme for the three reading functions?  If not, I suggest `ReadGem_v0_9_single_python`, `ReadGem_v0_9_single_pandas`, and `ReadGem_v0_9_single_cython`.

**Comment by kevinsa5** [comment]
(reposting with correct account, argh)

Merge conflict resolved.  A few questions:

1) Right now I've just copied the basic functionality test of `ReadGem_v0_9_single` for this.  Should I copy the other tests as well (empty/corrupt)?
2) For the `read_datafile(filename, method=...)` idea, does it make sense to have a wrapper function at the single datafile level, or would it be better to add that parameter to `ReadGem` and have it switch between the lower level functions directly?
3) Still need to rename `read_cython` to something better -- any preference for a naming scheme for the three reading functions?  If not, I suggest `ReadGem_v0_9_single_python`, `ReadGem_v0_9_single_pandas`, and `ReadGem_v0_9_single_cython`.

**Comment by ajakef** [comment]
How about restructuring so that  read_with_cython is called by ReadGem_v0_9_single as the first option in the "try" sequence? So it'll try the cython code first, then the fast pandas code, then the slow-but-reliable pandas code? That's what I meant in my last comment but I see it wasn't clear.

1. With the cython code being called by ReadGem_v0_9_single, it will be tested by existing tests. Is it necessary to make new tests?

2. I think method decisions should be made at the single-file level to ensure that the fastest workable read method will be used for each file. 

3. With read_with_cython now being called by ReadGem_v0_9_single, I don't think we need to make those function names similar to each other. Should non-user-callable functions like this one begin with underscores?

**Comment by kevinsa5** [comment]
> So it'll try the cython code first, then the fast pandas code, then the slow-but-reliable pandas code

Ok, got it.  Thanks for explaining, that approach sounds fine to me.

> With the cython code being called by ReadGem_v0_9_single, it will be tested by existing tests. Is it necessary to make new tests?

I still think it makes sense to test each of the three of the reader functions against each of the test files.  If the only tests were at the level of ReadGem_v0_9_single (with the try/except chaining like you described), I think you could only detect whether at least one of the three readers worked, so even if the cython code failed every time, ReadGem_v0_9_single tests would still pass if the pandas or python code worked.

> Should non-user-callable functions like this one begin with underscores?

If they aren't intended for users to call directly, yep.  Note that it's just a convention --  python doesn't assign any semantic meaning to an underscore prefix, but IDEs and such won't autocomplete them and most folks will be hesitant to use them in code.

File: gemlog/gemlog.py [hunk 1]
```diff
@@ -505,56 +505,133 @@ def MillisToTime(L):
...
+        try:
+            return reader(filename, offset)
+        except (EmptyRawFile, FileNotFoundError):
+            # If the file is definitely not going to work, exit early and
+            # re-raise the exception that caused the problem
+            raise
+        except Exception:
+            pass
+
+    raise CorruptRawFile(filename)
```
Comment (kevinsa5) [inline gemlog/gemlog.py:603]: Is this the correct error to raise if none of the readers can handle the file?

Comment (ajakef) [reply gemlog/gemlog.py:603]: Yep. Merging. Thanks again for the improvement!

**Pull request merged**